### PR TITLE
Add RabbitMQ Exchange

### LINKED ISSUE #34: Add RabbitMQ exchange
Add a rabbitmq exchange to notifiy other services when a new user is created.

### PR DESCRIPTION
**Pull request by ErikssonF**
Closes #34

### CODE REVIEW

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -10,3 +10,4 @@ spring.cloud.consul.discovery.register=true
 spring.cloud.consul.discovery.prefer-ip-address=true
 spring.cloud.consul.discovery.instance-id=${spring.application.name}:${spring.cloud.client.hostname}:${random.int[1,999999]}
 spring.cloud.consul.host=consul
+spring.rabbitmq.addresses=localhost:5672
```
Comment (Vimbayinashe) [inline src/main/resources/application.properties]:
```suggestion
spring.rabbitmq.host=rabbit
spring.rabbitmq.port=5672
```
The `application.properties` needs reference to the group's rabbit container instead of localhost because other services will not be able to reach localhost. Martin suggested that all services use the name **rabbit**. 

This can be tested by creating a local container named "rabbit" with the docker-compose file or by refering to localhost in one's `application-dev.properties`.

Comment (ErikssonF) [reply src/main/resources/application.properties]: My bad, I forgot to change from localhost. I'll add the container in our docker-compose as well

File: src/main/java/com/example/authservice/service/TokenService.java [hunk 2]
```diff
@@ -36,7 +37,7 @@ public Token generateToken(User user) {
                 .setIssuedAt(new Date(System.currentTimeMillis()))
                 .setExpiration(expiration)
                 .claim("roles", roles)
-                .signWith(key)
+                .signWith(SignatureAlgorithm.RS256,key)
```
Comment (Vimbayinashe) [inline src/main/java/com/example/authservice/service/TokenService.java:40]: The new method signature used here has been deprecated according to JwtBuilder documentation.

> Signs the constructed JWT using the specified algorithm with the specified key, producing a JWS.
> Deprecation Notice: Deprecated as of 0.10.0
> Use Keys.hmacShaKeyFor(bytes) to obtain the Key and then invoke signWith(Key)

Documentation states further that _"It is typically recommended to call the signWith(Key) instead for simplicity."_ so I suggested that we revert the changes made to this line.

File: src/main/java/com/example/authservice/messaging/CustomMessage.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@ToString
+public class CustomMessage {
+
+    private String messageId;
+    private String message;
+    private Date messageDate;
```
Comment (Vimbayinashe) [inline src/main/java/com/example/authservice/messaging/CustomMessage.java]:
```suggestion
    private String message;
```

We have been requested to publish a message with the new user's _"username"_ and therefore the other fields are not necessary unless UserService have specifically requested this.

Comment (Vimbayinashe) [reply src/main/java/com/example/authservice/messaging/CustomMessage.java]: Does rabbitTemplate require a unique id?

File: src/main/java/com/example/authservice/service/TokenService.java (hunk 2 above)
Comment (ErikssonF) [reply src/main/java/com/example/authservice/service/TokenService.java:40]: Alright, I couldn't run the app without this change, so it wasn't an intended change. I guess that I've used an old dependency somewhere.

File: src/main/java/com/example/authservice/messaging/MQConfig.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+@Configuration
+public class MQConfig {
+
+    public static final String QUEUE = "message_queue";
+    public static final String EXCHANGE = "message_exchange";
+    public static final String ROUTING_KEY = "message_exchange";
+
+    @Bean
+    public Queue queue(){
```
Comment (Vimbayinashe) [inline src/main/java/com/example/authservice/messaging/MQConfig.java]: Are we using the queue?

Comment (Vimbayinashe) [reply src/main/java/com/example/authservice/messaging/MQConfig.java]: It seems to only be used in MQConfig and so it can be deleted.

File: src/main/java/com/example/authservice/messaging/MQConfig.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        return new Queue(QUEUE);
+    }
+
+    @Bean
+    public TopicExchange exchange (){
+        return new TopicExchange(EXCHANGE);
+    }
+
+    @Bean
+    public Binding binding(Queue queue, TopicExchange exchange){
```
Comment (Vimbayinashe) [inline src/main/java/com/example/authservice/messaging/MQConfig.java]: If the queue is not being used then this binding method is not needed either.

File: src/main/java/com/example/authservice/messaging/MQConfig.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
+import org.springframework.amqp.support.converter.MessageConverter;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class MQConfig {
+
+    public static final String QUEUE = "message_queue";
```
Comment (Vimbayinashe) [inline src/main/java/com/example/authservice/messaging/MQConfig.java]:
```suggestion
```

File: src/main/java/com/example/authservice/messaging/MQConfig.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String EXCHANGE = "message_exchange";
+    public static final String ROUTING_KEY = "message_exchange";
```
Comment (Vimbayinashe) [inline src/main/java/com/example/authservice/messaging/MQConfig.java]: A descriptive or unique exchange name will be helpful for identifying this exchange amongst other exchanges and perhaps a routing key with a similar format to that in both the handouts and Martin's example is also a good idea.
https://github.com/kappsegla/JWTAuthWithGatewayExample/commit/e346474fbe50871657f6a302ce22c2cc7727cd87

File: pom.xml [hunk 8]
```diff
@@ -55,27 +55,68 @@
...
+        <dependency>
+            <groupId>com.rabbitmq</groupId>
+            <artifactId>amqp-client</artifactId>
+            <version>5.14.2</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.amqp</groupId>
+            <artifactId>spring-amqp</artifactId>
+            <version>2.4.4</version>
+        </dependency>
```
Comment (Vimbayinashe) [inline pom.xml]: These two dependencies do not seem to be used in the implementation. `spring-boot-starter-amqp` is likely sufficient

File: src/main/java/com/example/authservice/messaging/CustomMessage.java (hunk 3 above)
Comment (ErikssonF) [reply src/main/java/com/example/authservice/messaging/CustomMessage.java]: I'm not sure about the id, I guess we don't have any functionality that uses it? Are we going to need it for anything in the final project?

File: src/main/java/com/example/authservice/messaging/MQConfig.java (hunk 4 above)
Comment (ErikssonF) [reply src/main/java/com/example/authservice/messaging/MQConfig.java]: If we don't need a message queue for this service then we can remove it. 
It was added in the example I followed and I'm not certain how our service is going to be used when we put all services together, so I think I need to figure that out first before we remove it.

File: src/main/java/com/example/authservice/messaging/MessagePublisher.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.Date;
+import java.util.UUID;
+
+@RestController
```
Comment (Vimbayinashe) [inline src/main/java/com/example/authservice/messaging/MessagePublisher.java]: This should be a regular class without `RestController` and we should also remove the` @RequestBody ` annotation on line 18.

According to Baeldung:
> Spring 4.0 introduced the @RestController annotation in order to simplify the creation of RESTful web services.

File: src/main/java/com/example/authservice/messaging/MessagePublisher.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+public class MessagePublisher {
+
+    @Autowired
```
Comment (Vimbayinashe) [inline src/main/java/com/example/authservice/messaging/MessagePublisher.java]: Constructor injection is highly preferable instead of use of the `@Autowired` annotation.

[A good explanation why one should use constructor injection](https://stackoverflow.com/a/40620318)
[An article on constructor injection](https://reflectoring.io/constructor-injection/)

File: src/main/java/com/example/authservice/messaging/CustomMessage.java (hunk 3 above)
Comment (Vimbayinashe) [reply src/main/java/com/example/authservice/messaging/CustomMessage.java]: The only thing that they requested from us was the email address. You can test and if the exchange still works then we can remove it.

File: pom.xml (hunk 8 above)
Comment (Vimbayinashe) [reply pom.xml]:
```suggestion
```
I forgot to highlight them, I meant amqp-client and spring-amqp. The application is running and publishing messages to the exchange without them.

File: src/main/java/com/example/authservice/messaging/MQConfig.java (hunk 4 above)
Comment (Vimbayinashe) [reply src/main/java/com/example/authservice/messaging/MQConfig.java]: According to RabbitMQ's documentation
> The core idea in the messaging model in RabbitMQ is that the producer never sends any messages directly to a queue. Actually, quite often the producer doesn't even know if a message will be delivered to any queue at all.

you can test to comment out the queue's code and a new user registration can still be published.

**Comment by Vimbayinashe** [comment]
This is working well! Great work! I have left comments on some issues.

File: pom.xml [hunk 11]
```diff
@@ -55,27 +55,68 @@
             <artifactId>hibernate-core</artifactId>
             <version>5.6.8.Final</version>
         </dependency>
+        <dependency>
+            <groupId>com.auth0</groupId>
+            <artifactId>java-jwt</artifactId>
+            <version>3.10.3</version>
+        </dependency>
```
Comment (Vimbayinashe) [inline pom.xml]:
```suggestion
```

This dependency does not seem to be used and removing it does not throw an error

Comment (ErikssonF) [reply pom.xml]: I'm not sure why these followed onto this branch, but it wasn't an intended addition for messaging.

File: pom.xml [hunk 12]
```diff
@@ -55,27 +55,68 @@
...
+        <dependency>
+            <groupId>io.github.cdimascio</groupId>
+            <artifactId>java-dotenv</artifactId>
+            <version>5.2.2</version>
+        </dependency>
```
Comment (Vimbayinashe) [inline pom.xml]:
```suggestion
        <dependency>
            <groupId>io.github.cdimascio</groupId>
            <artifactId>dotenv-java</artifactId>
            <version>2.2.3</version>
        </dependency>
```
Please use this dependency instead according to their [documentation](https://github.com/cdimascio/java-dotenv#why)

Comment (ErikssonF) [reply pom.xml]: Sure!

File: src/main/java/com/example/authservice/service/TokenService.java (hunk 2 above)
Comment (Vimbayinashe) [reply src/main/java/com/example/authservice/service/TokenService.java:40]: It seems that the upgrade in jwt dependency versions allows us to now use `.signWith(key)`. Great problem solving 😉

File: pom.xml (hunk 8 above)
Comment (ErikssonF) [reply pom.xml]: Resolved this with latest push to branch

**Review by Vimbayinashe — APPROVED** [review]
